frontend/exporter/src/id_table.rs

```diff
@@ -0,0 +1,375 @@
+        Value::Ty(value)
+    }
+    fn from_types(t: &Value) -> Option<Arc<Self>> {
+        match t {
+            Value::Ty(value) => Some(value.clone()),
+        }
+    }
+}
+
+/// A value encodable in the enum `Value`
+pub trait SupportedType: std::fmt::Debug {
+    fn to_types(value: Arc<Self>) -> Value;
+    fn from_types(t: &Value) -> Option<Arc<Self>>;
+}
+
+impl Session {
+    fn fresh_id(&mut self) -> Id {
+        let id = self.next_id.id;
+        self.next_id.id += 1;
+        Id { id }
+    }
+}
+
+impl<T: Sync + Send + Clone + 'static + SupportedType> Node<T> {
+    pub fn new(value: T, session: &mut Session) -> Self {
+        let id = session.fresh_id();
+        let kind = Arc::new(value);
+        session.table.insert(id.clone(), kind.clone());
+        Self { id, value: kind }
+    }
+    pub fn inner(&self) -> Arc<T> {
+        self.value.clone()
+    }
+
+    pub fn kind(&self) -> &T {
+        self.value.as_ref()
+    }
+}
+
+/// Wrapper for a type `T` that creates a bundle containing both a ID
+/// table and a value `T`. That value may contains `Node` values
+/// inside it. Serializing `WithTable<T>` will serialize IDs only,
+/// skipping values. Deserialization of a `WithTable<T>` will
+/// automatically use the table and IDs to reconstruct skipped values.
+#[derive(Serialize, Debug)]
+pub struct WithTable<T> {
+    table: Table,
+    value: T,
+}
+
+/// The state used for deserialization: a table.
+static DESERIALIZATION_STATE: LazyLock<Mutex<Table>> =
+    LazyLock::new(|| Mutex::new(Table::default()));
+static DESERIALIZATION_STATE_LOCK: LazyLock<Mutex<()>> = LazyLock::new(|| Mutex::new(()));
+
+/// The mode of serialization: should `Node<T>` ship values of type `T` or not?
+static SERIALIZATION_MODE_USE_IDS: std::sync::atomic::AtomicBool =
+    std::sync::atomic::AtomicBool::new(false);
+
+fn serialize_use_id() -> bool {
+    SERIALIZATION_MODE_USE_IDS.load(Ordering::Relaxed)
+}
+
+impl<T> WithTable<T> {
+    /// Runs `f` with a `WithTable<T>` created out of `map` and
+    /// `value`. Any serialization of values of type `Node<_>` will
+    /// skip the field `value`.
+    pub fn run<R>(map: Table, value: T, f: impl FnOnce(&Self) -> R) -> R {
+        if serialize_use_id() {
+            panic!("CACHE_MAP_LOCK: only one WithTable serialization can occur at a time (nesting is forbidden)")
+        }
+        SERIALIZATION_MODE_USE_IDS.store(true, Ordering::Relaxed);
+        let result = f(&Self { table: map, value });
+        SERIALIZATION_MODE_USE_IDS.store(false, Ordering::Relaxed);
+        result
+    }
+    pub fn destruct(self) -> (T, Table) {
+        let Self { value, table: map } = self;
+        (value, map)
+    }
+}
+
+/// Helper function that makes sure no nested deserializations occur.
+fn full_id_deserialization<T>(f: impl FnOnce() -> T) -> T {
+    let _lock = *DESERIALIZATION_STATE_LOCK.lock().expect("CACHE_MAP_LOCK: only one WithTable deserialization can occur at a time (nesting is forbidden)");
```

Comment:
Should you dereference here? Doesn't that copy the `()` and drop the guard? I might be missing a level

Reply:
Uuh, thanks! was doing something else with that before, now that makes it useless!
I will drop that and add a type annotation to make it clear

Reply:
Also, `lock` blocks waiting for someone else holding the lock, so this will deadlock in case of nested deserializations. You want `try_lock` instead. (the `Result` that you call `expect` on is for lock poisoning)

Reply:
ah yes right, thank you, that's a good catch, this would have been pretty horrible to debug!